File: part_7/main.c
#include <inttypes.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *toBinaryString(long long int number);

int main(void) {
    FILE *file = fopen("hyperskill-dataset-117460134.txt", "r");
    if (file == NULL) {
        perror("Error opening the file");
        return EXIT_FAILURE;
    }

    char number[64] = {0};
    fgets(number, sizeof(number), file);
    fclose(file);
    number[sizeof(number) - 1] = '\0';

	// Remove newline character
    for (size_t i = 0; i < sizeof(number) / sizeof(number[0]); i++) {
        if (number[i] == '\n') {
            number[i] = '\0';
            break;
        }
    }

    long long int phoneNumber = strtoll(number, NULL, 10);
    long long int uintMax = UINT32_MAX + 1LL; // Exercise considers 2^32 not 2^32-1.
    int numberOfOverflows = 0;

    while (phoneNumber >= uintMax) {
        phoneNumber -= uintMax;
        numberOfOverflows++;
    }

    // Format example: 1,11001010001001111000111011001100
    char *binaryPhoneNumber = toBinaryString(phoneNumber);
    printf("%d,%s\n", numberOfOverflows, binaryPhoneNumber);
    free(binaryPhoneNumber);

    return 0;
}

/*
 * Convert an integer to a binary string
 */
char *toBinaryString(long long int number) {
    // Handle special case for 0 (zero)
    if (number == 0) {
        char *result = malloc(2);
        if (result == NULL) {
            perror("Failed to allocate memory");
            return NULL;
        }

        strcpy(result, "0");
        return result;
    }

    // Calculate number of bits needed
    long long int temp = number;
    int bitCount = 0;
    while (temp != 0) {
        bitCount++;
        temp >>= 1;
    }

    // Allocate memory for binary string (+1 for null terminator)
    char *binary = malloc(bitCount + 1);
    if (binary == NULL) {
        perror("Failed to allocate memory");
        return NULL;
    }

    // Convert to binary
    int index = bitCount - 1;
    temp = number;

    while (temp != 0) {
        binary[index] = (temp & 1) ? '1' : '0';
        temp >>= 1;
        index--;
    }

    binary[bitCount] = '\0'; // Null terminate the string
    return binary; // Return pointer to memory allocated string
}

File: part_7/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class Main {

    public static String readFile() {
        String filename = "hyperskill-dataset-117458807.txt";
        Path filepath = Path.of(filename);

        try {
            return Files.readString(filepath).trim();
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            return "";
        }
    }

    // 1,11001010001001111000111011001100
    public static void main(String[] args) {
        long phoneNumber = Long.parseLong(readFile());
        long uintMax = ((long) Integer.MAX_VALUE) * 2 + 2; // Exercise considers 2^32 not 2^32-1
        System.out.println(uintMax);
        int overflows = 0;

        while (phoneNumber >= uintMax) {
            phoneNumber -= uintMax;
            overflows++;
        }

        String binaryPhoneNumber = Long.toBinaryString(phoneNumber);

        System.out.printf("%d,%s\n", overflows, binaryPhoneNumber);
    }
}
